Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm

```diff
@@ -239,6 +249,71 @@ void requestLinkDecorationFilteringData(LinkFilteringRulesCallback&& callback)
     }];
 }
 
+StorageAccessUserAgentStringQuirkController& StorageAccessUserAgentStringQuirkController::shared()
+{
+    static StorageAccessUserAgentStringQuirkController* sharedInstance = new StorageAccessUserAgentStringQuirkController;
+    return *sharedInstance;
+}
+
+Ref<StorageAccessUserAgentStringQuirkObserver> StorageAccessUserAgentStringQuirkController::observeUpdates(Function<void()>&& callback)
+{
+    if (!m_notificationListener) {
+        m_notificationListener = adoptNS([WKWebPrivacyNotificationListener new]);
+        [m_notificationListener listenForStorageAccessUserAgentStringQuirkChanges:^{
+            updateQuirks([this] {
+                for (auto& observer : m_observers)
+                    observer.invokeCallback();
+            });
+        }];
+    }
+    auto observer = StorageAccessUserAgentStringQuirkObserver::create(WTFMove(callback));
+    m_observers.add(observer.get());
+    return observer;
+}
+
+void StorageAccessUserAgentStringQuirkController::setCachedQuirks(Vector<WebCore::UserAgentStringQuirk>&& quirks)
+{
+    m_cachedQuirks = WTFMove(quirks);
+    m_cachedQuirks.shrinkToFit();
+}
+
+void StorageAccessUserAgentStringQuirkController::setCachedQuirksForTesting(Vector<WebCore::UserAgentStringQuirk>&& quirks)
+{
+    setCachedQuirks(WTFMove(quirks));
+}
+
+void StorageAccessUserAgentStringQuirkController::updateQuirks(CompletionHandler<void()>&& callback)
+{
+    if (!WebKit::canUseWebPrivacyFramework()) {
+        callback();
+        return;
+    }
+
+    static NeverDestroyed<Vector<CompletionHandler<void()>, 1>> lookupCallbacks;
+    lookupCallbacks->append(WTFMove(callback));
+    if (lookupCallbacks->size() > 1)
+        return;
+
+    auto options = adoptNS([PAL::allocWPResourceRequestOptionsInstance() init]);
+    [options setAfterUpdates:NO];
+
+    [[PAL::getWPResourcesClass() sharedInstance] requestStorageAccessUserAgentStringQuirksData:options.get() completionHandler:^(WPStorageAccessUserAgentStringQuirksData *data, NSError *error) {
+        Vector<WebCore::UserAgentStringQuirk> result;
```

Comment:
Any reason we're using a `Vector<WebCore::UserAgentStringQuirk>` in this class? Why can't we store a `HashMap<RegistrableDomain, String /* user agent */>` here, since it seems to be what we need eventually? Why do we need the UserAgentStringQuirk struct at all?